src/compiler/scala/tools/nsc/typechecker/PostTyperChecks.scala

```diff
@@ -0,0 +1,65 @@
+/* NSC -- new Scala compiler
+ * Copyright 2005-2013 LAMP/EPFL
+ * @author  Martin Odersky
+ */
+
+package scala.tools.nsc
+package typechecker
+
+import scala.reflect.internal.util.StatisticsStatics
+
+/** The main attribution phase.
+ */
+trait PostTyperChecks extends AnyRef
+//            with Contexts
+//            with Namers
+//            with Typers
+//            with Infer
+//            with Implicits
+//            with EtaExpansion
+//            with SyntheticMethods
+//            with Unapplies
+//            with Macros
+//            with NamesDefaults
+//            with TypeDiagnostics
+//            with ContextErrors
+//            with StdAttachments
+//            with AnalyzerPlugins
+{
+  val global : Global
+
+  import global._
+
+  object postTyperFactory extends {
+    val global: PostTyperChecks.this.global.type = PostTyperChecks.this.global
+  } with SubComponent {
+    import global.statistics
+    val phaseName = "postTyperChecks"
+    val runsAfter = List[String]("typer")
+    val runsRightAfter = None
+    def newPhase(_prev: Phase): StdPhase = new StdPhase(_prev) {
+      override def run(): Unit = {
+        val start = if (StatisticsStatics.areSomeColdStatsEnabled) statistics.startTimer(statistics.postTyperNanos) else null
+        global.echoPhaseSummary(this)
+        for (unit <- currentRun.units) {
+          applyPhase(unit)
+        }
+        if (StatisticsStatics.areSomeColdStatsEnabled) statistics.stopTimer(statistics.postTyperNanos, start)
+      }
+      def apply(unit: CompilationUnit): Unit = {
+        try {
+          if (global.settings.Yrangepos) global.validatePositions(unit.body)
+          for (workItem <- unit.toCheck) workItem()
```

Comment:
Relatedly, this could cause a leak in the presentation compiler if a long-lived `CompilationUnit` is typechecked repeatedly, as it will accumulate actions in `toCheck` that are not cleared.

Reply:
will move this back

Reply:
and consider later